docs: add usage guide

alexeagle opened the pull request.

JiaLiPassion commented on CONTRIBUTING.md:
```diff
@@ -31,11 +31,13 @@ directory:
 
-This means that any usage of `@aspect_rules_js` on your system will point to this folder.
+This means that any usage of `@aspect_rules_js` on your system will point to this folder, if you
+pass the `--config=override` flag to Bazel.
```
Should we name the name of this config to `dev` or `local`?

JiaLiPassion commented on CONTRIBUTING.md:
```diff
@@ -31,11 +31,13 @@ directory:
+(We don't want to enable this for all builds, since you'll later forget this was done and wonder
+why you don't reproduce the same behavior as others.)
 
 ## Releasing
```
Also not related to this PR, is there any way to do a local `test` before commit and push?

alexeagle replied:
sure, `bazel test ...` like usual.

JiaLiPassion commented on docs/README.md:
```diff
@@ -0,0 +1,320 @@
+<https://github.com/aspect-build/rules_js/releases>
+copy the WORKSPACE snippet into your `WORKSPACE` file.
+
+Note that [bzlmod] is experimentally supported as well.
```
Should we add a `snippet` to describe how to use `bzlmod` ?

thesayyn replied:
maybe link to `e2e/bzlmod` for the full example.

JiaLiPassion commented on docs/README.md:
```diff
@@ -0,0 +1,320 @@
+
+**API docs:**
+
+-   Installing and choosing the version of Node.js:
```
In my understanding, the `nodejs` toolchain is installed in the `WORKSPACE` snippet above, we don't need to `install` nodejs ourselves. Is that correct?

alexeagle replied:
that's right. I'll just label this one "choosing the version..."

JiaLiPassion commented on docs/README.md:
```diff
@@ -0,0 +1,320 @@
+
+Next, you'll typically use `npm_translate_lock` to translate the Yaml file to Starlark, which Bazel extensions understand.
+The `WORKSPACE` snippet you pasted above already contains this code.
+
```
Should we post two screenshots or code here to display the `pnpm lock` and the `Starlark` version side by side, it may be eaiser for the user to understand.

alexeagle replied:
I don't think users would understand the content of either file actually. They're both opaque blobs you never read, and GitHub and other code review tools don't even show you the changes. I did add a section here about how you could check in the generated file though, rather than compute on-the-fly. Do you think it's understandable?

JiaLiPassion commented on docs/README.md:
```diff
@@ -0,0 +1,320 @@
+
+Technically, we run a port of pnpm rather than pnpm itself. The reasoning is as follows:
+
+1. You don't need to install pnpm on your machine to build and test with Bazel.
+1. We re-use pnpm's resolver, by consuming the `pnpm-lock.yaml` file it produces.
+1. We use Bazel's downloader API to fetch package tarballs and extract them to external repositories.
+   To modify the URLs Bazel uses to download packages (for example, to fetch from Artifactory), read
+   <https://blog.aspect.dev/configuring-bazels-downloader>.
+1. We re-use the [`@pnpm/lifecycle`](https://www.npmjs.com/package/@pnpm/lifecycle) package to perform postinstall steps.
+   (These run as cacheable Bazel actions.)
+1. Finally, you link the `node_modules` tree by adding a `npm_link_package` or `npm_link_all_packages` in your `BUILD` file,
+   which populates a tree under `bazel-bin/[path/to/package]/node_modules`.
```
I believe this is the most important part of the rulejs core design logic, maybe we can add more information with sequence to describe the flow with more details in the future PR.

alexeagle replied:
added a section below about how we use the bazel-bin folder for srcs, genfiles, and deps

JiaLiPassion commented on docs/README.md:
```diff
@@ -0,0 +1,320 @@
+We recommend producing Markdown output, and checking those `.md` files into your source repository.
+This makes it easy to browse them at the same revision as the sources.
+
+You'll need to create `bzl_library` targets for your starlark files.
```
Use `Starlark` (Capital first letter) to keep consistant in the doc.

JiaLiPassion commented on docs/README.md:
````diff
@@ -0,0 +1,320 @@
+)
+
+update_docs(name = "docs")
+```
````
Do we have a workable example in the `bazel_examples` repo? If we do have, we can add a link here.

JiaLiPassion commented on docs/js_library.md:
```diff
@@ -1,11 +1,25 @@
 <!-- Generated with Stardoc: http://skydoc.bazel.build -->
 
-Expose some files with DeclarationInfo, like filegroup but can be a dep of ts_project.
+`js_library` is similar to [`filegroup`](https://docs.bazel.build/versions/main/be/general.html#filegroup); there are no Bazel actions to run.
```
Just a question, is there any differences between `js_library` and `filegroup` rule?

alexeagle commented:
> 1. Tools installation.
Added the bazelisk installation note. Bazelisk installs Bazel, so there's no need to mention that, nor any version info.

> 2. Since we are using the 5.x of Bazel in the bazelrc, should we also update the pre-commit-buildifier version?
Sure, updated to latest in .pre-commit-config.yaml. However if you got some version skew, that seems like a bug - pre-commit should install the version we ask for regardless what you setup on your machine...
